README.md
node-psi-server README
======================

This document provides a short description of
[node-psi-server](http://github.com/jansc/node-psi-server/).
`node-psi-server` is a Topic Maps application.

Dependencies
------------

Make sure that you've installed these two applications on your machine:
 - [node.js](http://nodejs.org) - The NodeJS engine
 - [Connect](http://github.com/senchalabs/connect) -  Middleware for NodeJS.

The following external libraries have been included in the node-psi-server
package:
 - [tmjs](http://github.com/jansc/tmjs) - an Open Source Topic Maps engine.
 - [Mu](http://github.com/raycmorgan/Mu) - an asynchroneous template engine
   based on mustache.js.
 - [opts.js](http://bitbucket.org/mazzarell/js-opts/opts.js) - a command
   line options parser.


Examples
--------

You can use the Opera topic map from the examples catalog for testing.
To see which PSIs are included, run the following command:

  node server.js --list examples/opera.jtm psi.ontopia.net


To start the web server run

  node server.js --config ./config-sample.jtm examples/opera.jtm psi.ontopia.net 
 
 * config-sample.jtm is the configuration file,
 * examples/opera.jtm is the topic map that you want to serve PSIs for
 * psi.ontopia.net is the server name

Now you should be able to open e.g. http://localhost:8000/person or 
"http://localhost:8000/city/380-firenze" in your web browser.


Limitations
-----------

Besides all bugs, node-psi-server has two main limitations:

1. No support of PSIs with anchor.
2. Only topic maps in JTM format are supported. 

Detailed explanations:

'#' in PSIs is not supported. This is because the anchor part is not sent to
the server. If you e.g. enter 

  http://psi.example.com/foo#bar

the requested URL on the server is

  http://psi.example.com/foo

The is could be solved by creating a special index for all PSIs containing
a # symbol, but currently this feature is not implemented.

Due to a limitation in tmjs, only JTM topic maps can be imported. However,
you can use [mappify.org](http://ws.mappify.org) to convert any topic map
to JTM. [Maiana](http://maiana.topicmaplab.de) and
[Ontopia](http://ontopia.net) also support export of topic maps in JTM.



HISTORY
== 0.1.0 2010-09-28

* First public release


lib/tools.js
// node-psi-server, version 0.0.1
// http://github.com/jansc/node-psi-server
// Copyright (c) 2010 Jan Schreiber <jans@ravn.no>
// Licensed under the MIT-License.

/*jslint browser: true, devel: true, onevar: true, undef: true,
  nomen: false, eqeqeq: true, plusplus: true, bitwise: true, regexp: true,
    newcap: true, immed: true, indent: 4 */
/*global exports, require, startTimestamp*/

var util = require('util'),
    tmjs = require('../vendor/tm'),
    fs = require('fs'),
    defaultNameType = null,
    exporttmid = 0, // id of the exported json topic map
    configPrefix = 'http://psi.semanticheadache.com/node-psi-server/',
    configLocators = {
        CONFIGURATION_VALUE: configPrefix + 'configuration-value',
        LOGFILE: configPrefix + 'logfile',
        PORT: configPrefix + 'port',
        ROLE_TYPE: configPrefix + 'type',
        TYPE_IS_PUBLISHED: configPrefix + 'type-is-published',
        PUBLISHER_FALLBACK: configPrefix + 'publisher-fallback',
        PSI_STATUS_FALLBACK: configPrefix + 'psi-status',
        CREATED_FALLBACK: configPrefix + 'created-fallback',
        DEBUG_LEVEL: configPrefix + 'debug-level'
    };

exports.getTopicMap = function (iri) {
    var factory, tmsys, tmid;
    // Create a topic map and import the topic map file.
    factory = tmjs.TM.TopicMapSystemFactory.newInstance();
    factory.setProperty('com.semanticheadache.tmjs.backend', 'memory');
    tmsys = factory.newTopicMapSystem();
    tmid = tmsys.createLocator(iri);
    return tmsys.createTopicMap(tmid);
};


// Copies an source_occ to target_topic. If types or scoping topics
// don't exist in taret_topic's topic map, stubs for there topics are
// copied as well.
exports.copyOccurrence = function (sourceOcc, targetTopic) {
    var targetOcc,
        targetType;
    targetType = exports.createTopicStub(sourceOcc.getType());
    // Should we use createLocator if value is a locator?
    targetOcc = targetTopic.createOccurrence(targetType, sourceOcc.getValue(),
                                            sourceOcc.getDatatype());
    exports.copyScope(sourceOcc, targetOcc);
    exports.copyReifier(sourceOcc, targetOcc);
};

exports.copyAssociation = function (sourceAssoc, targetTM) {
    var i, targetAssoc,
        targetType,
        sourceScope, targetScope = [],
        sourceRoles, targetRole;
    targetType = exports.createTopicStub(sourceAssoc.getType(), targetTM);
    sourceScope = sourceAssoc.getScope();
    for (i = 0; i < sourceScope.length; i += 1) {
        targetScope.push(exports.createTopicStub(sourceScope[i], targetTM));
    }
    targetAssoc = targetTM.createAssociation(targetType);
    exports.copyScope(sourceAssoc, targetAssoc);
    exports.copyReifier(sourceAssoc, targetAssoc);

    sourceRoles = sourceAssoc.getRoles();
    for (i = 0; i < sourceRoles.length; i += 1) {
        targetAssoc.createRole(
            exports.createTopicStub(sourceRoles[i].getType(), targetTM),
            exports.createTopicStub(sourceRoles[i].getPlayer(), targetTM)
        );
        exports.copyReifier(sourceRoles[i], targetRole);
    }
    return targetAssoc;
};

exports.copyScope = function (source, target) {
    var i, sourceScope;
    sourceScope = source.getScope();
    for (i = 0; i < sourceScope.length; i += 1) {
        target.addTheme(exports.createTopicStub(sourceScope[i], target.getTopicMap()));
    }
};

exports.copyReifier = function (source, target) {
    // TODO: Implement
};

// Checks if a topic with one of the identifiers of sourceTopic
// exists in targetTM. If not, a new topic is created, and all identifiers
// are being copied.
// Returns the existing or created topic in targetTM
exports.createTopicStub = function (sourceTopic, targetTM) {
    var i, topic = null, psis, iis;
    // Copy PSIs
    psis = sourceTopic.getSubjectIdentifiers();
    for (i = 0; i < psis.length; i += 1) {
        if ((topic = targetTM.getTopicBySubjectIdentifier(
            targetTM.createLocator(psis[i].getReference())))) {
            break;
        }
    }
    // Did not find topic by subject identifier
    // Try subject locators
    if (!topic) {
        psis = sourceTopic.getSubjectIdentifiers();
        for (i = 0; i < psis.length; i += 1) {
            if ((topic = targetTM.getTopicBySubjectIdentifier(
                targetTM.createLocator(psis[i].getReference())))) {
                break;
            }
        }
    }
    // Still no topic... so we have to create a new topic
    if (!topic) {
        iis = sourceTopic.getSubjectIdentifiers();
        if (iis.length > 0) {
            topic = targetTM.createTopicBySubjectIdentifier(
                targetTM.createLocator(iis[0].getReference()));
        } else {
            iis = sourceTopic.getSubjectLocators();
            if (iis.length > 0) {
                return iis[0].getReference();
            } else {
                topic = targetTM.createTopic();
            }
        }
    }
    return topic;
};


// Creates a new empty topic map and copies all relevant information
// about topic to the new topic map. The new topic map is returned.
// TODO: Should be part of tmjs' CopyUtils.
exports.exportTopic = function (tm, topic, exportTypes) {
    var exporttm, i, j, psis, names, t, occs, roles, type;
    exporttmid += 1;
    exporttm = exports.getTopicMap('http://semanticheadache.com/node-psi-server/export/' + exporttmid);
    psis = topic.getSubjectIdentifiers();
    t = exporttm.createTopicBySubjectIdentifier(exporttm.createLocator(psis[0].getReference()));
    // Copy all other PSIs
    for (i = 1; i < psis.length; i += 1) {
        t.addSubjectIdentifier(exporttm.createLocator(psis[i].getReference())); 
    }

    for (type in exportTypes) {
        if (exportTypes.hasOwnProperty(type)) {
            occs = topic.getOccurrences(exportTypes[type]);
            for (j = 0; j < occs.length; j += 1) {
                // TODO: Could be TM.CopyHelper.copyOccurrence();
                exports.copyOccurrence(occs[j], t);
            }
        }
    }

    // Not the most performant way. Refactor!
    for (type in exportTypes) {
        if (exportTypes.hasOwnProperty(type)) {
            roles = topic.getRolesPlayed();
            for (j = 0; j < roles.length; j += 1) {
                if (roles[j].getParent().getType().equals(exportTypes[type])) {
                    exports.copyAssociation(roles[j].getParent(), exporttm);
                }
            }
        }
    }

    // TODO: Copy subject locators
    names = topic.getNames();
    for (i = 0; i < names.length; i += 1) {
        t.createName(names[i].getValue()); // TODO: Copy scope and scoping topics
    }

    return exporttm;
};

// Takes a topic object as parameter and returns a textual label
// that represents that topic. If anchor is set, a HTML-link
// is returned. The URL is a 
// options can contain one of the following:
// * anchor (boolean). If true, getTopicLabel tries to wrap the the label
//   in an HTML link
// * servername (String). Servername of the local server. Use this in combination
//   with the anchor. If a URL has the same domain name as servername, a relative
//   URL is generated
exports.getTopicLabel = function (topic, options) {
    var label = null, foundDefaultName = false, minScopeLen = 1000,
        isName = false, names, i, tm, url;
    options = options || {};
    tm = topic.getParent();
    if (!defaultNameType) {
        defaultNameType = tm.getTopicBySubjectIdentifier(
            tm.createLocator(tmjs.TM.TMDM.TOPIC_NAME));
    }

    if ((names = topic.getNames()) && names.length > 0) {
        for (i = 0; i < names.length; i += 1) {
            if (defaultNameType && names[i].getType().equals(defaultNameType)) {
                // If several default names are available,
                // pick the one with fewest scoping topics
                if (label && foundDefaultName &&
                    names[i].getScope().length >= minScopeLen) {
                    continue;
                }
                label = names[i];
                foundDefaultName = true;
                minScopeLen = names[i].getScope().length;
            } else {
                if (!label) {
                    // Better than nothing
                    label = names[i];
                }
            }
        }
        label = label.getValue();
        isName = true;
    } else if (topic.getSubjectIdentifiers().length > 0) {
        label = topic.getSubjectIdentifiers()[0].getReference();
    } else if (topic.getItemIdentifiers().length > 0) {
        label = topic.getItemIdentifiers()[0].getReference();
    }
    if (options.anchor) {
        url = exports.getTopicURL(topic);
        if (url) {
            // TODO check if URL belongs to servername!
            label = '<a href="' + url + '">' + label + '</a>';
        }
    }

    return {label: label, isName: isName};
};

exports.getTopicURL = function (topic) {
    var iis;
    iis = topic.getSubjectIdentifiers();
    if (iis.length > 0) {
        return iis[0].getReference();
    }
    iis = topic.getSubjectLocators();
    if (iis.length > 0) {
        return iis[0].getReference();
    }
    return null;
};


// Parses the configuration topic map
exports.parseConfigTM = function (configTM, opts, config) {
    var filename = opts.get('c'), values, getConfigValue, data,
        reader, tmData, typeIsPublished, type, index, i, j,
        associations, roles, topic, identifiers = [], tmp;
    // If the user specified a configuration file, read and parse it
    if (opts.get('c')) {
        getConfigValue = function (loc_str) {
            var tmp, valueType;
            valueType = configTM.getTopicBySubjectIdentifier(configTM.createLocator(
                configLocators.CONFIGURATION_VALUE));

            if (valueType) {        
                tmp = configTM.getTopicBySubjectIdentifier(configTM.createLocator(
                    loc_str));
                if (tmp) {
                    values = tmp.getOccurrences(valueType);
                    if (values.length) {
                        return values[0].getValue();
                    }
                }
            } else {
                util.log("Could not find " + configLocators.CONFIGURATION_VALUE);
            }
            return null;
        };
        util.log("Reading configuration file '" + filename + "'.");
        data = fs.readFileSync(filename);
        tmData = JSON.parse(data);
        reader = new tmjs.TM.JTM.Reader(configTM);
        reader.fromObject(tmData);

        config.port = getConfigValue(configLocators.PORT);
        config.publisherFallback = getConfigValue(configLocators.PUBLISHER_FALLBACK);
        config.psiStatusFallback = getConfigValue(configLocators.PSI_STATUS_FALLBACK);
        config.createdFallback = getConfigValue(configLocators.CREATED_FALLBACK);
        config.debugLevel = getConfigValue(configLocators.DEBUG_LEVEL);
        config.logFile = getConfigValue(configLocators.LOGFILE);

        // get types to be published
        typeIsPublished = configTM.getTopicBySubjectIdentifier(
            configTM.createLocator(configLocators.TYPE_IS_PUBLISHED));
        type = configTM.getTopicBySubjectIdentifier(
            configTM.createLocator(configLocators.ROLE_TYPE));

        // make sure that both topics exists, if not we can drop that part
        if (!type || !typeIsPublished) {
            return;
        }    
        index = configTM.getIndex('TypeInstanceIndex');
        associations = index.getAssociations(typeIsPublished);
        for (i = 0; i < associations.length; i += 1) {
            roles = associations[i].getRoles();
            // assume unary associations
            if (roles.length !== 1) {
                continue;
            }
            // should we check the type of the role?
            topic = roles[0].getPlayer();
            tmp = topic.getSubjectIdentifiers();
            for (j = 0; j < tmp.length; j += 1) {
                identifiers.push(tmp[j].getReference());
            }
            tmp = topic.getSubjectLocators();
            for (j = 0; j < tmp.length; j += 1) {
                identifiers.push(tmp[j].getReference());
            }
        }
    }
    return identifiers;
};

// Prefetches all topic types we'd like to export
exports.prefetchExportTypes = function (tm, exportIdentifiers) {
    var topic, exportTypes = {}, i;
    if (exportIdentifiers.length > 0) {
        util.log('Occurrence types and association types to include on PSI pages:');
        for (i = 0; i < exportIdentifiers.length; i += 1) {
            util.log(' * ' + exportIdentifiers[i]);
        }
        util.log('Prefetching export types:');
        for (i = 0; i < exportIdentifiers.length; i += 1) {
            topic = tm.getTopicBySubjectIdentifier(tm.createLocator(exportIdentifiers[i]));
            if (topic) {
                util.log(' * found ' + exportIdentifiers[i]);
                exportTypes[exportIdentifiers[i]] = topic;
            }
        }
    }
    return exportTypes;
};



server.js
// node-psi-server, version 0.0.1
// http://github.com/jansc/node-psi-server
// Copyright (c) 2010 Jan Schreiber <jans@ravn.no>
// Licensed under the MIT-License.
//
// Permission is hereby granted, free of charge, to any person
// obtaining a copy of this software and associated documentation
// files (the "Software"), to deal in the Software without
// restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following
// conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
// OTHER DEALINGS IN THE SOFTWARE.

/*jslint browser: true, devel: true, onevar: true, undef: true,
  nomen: false, eqeqeq: true, plusplus: true, bitwise: true, regexp: true,
    newcap: true, immed: true, indent: 4 */
/*global require, __dirname*/

(function () {
    var util = require('util'),
        tmjs = require('./vendor/tm'),
        fs = require('fs'),
        opts = require('./vendor/opts'),
        Connect = require('connect'),
        Mu = require('./vendor/mu'),
        url = require('url'),
        tools = require('./lib/tools'),
        exportIdentifiers,
        tm, // The main topic map object
        options,
        cmdArguments,
        debug, // debug command line option, boolean
        tmFile, // Topic map file command line argument
        //logFile, // Log file command line option
        logStream, // Log file stream
        servername, // default host value
        showList, // boolean. true if command line option for listing all PSIs is set
        showJTM, // function that exports information about the referred topic to
        showHTML, // function that exports information about the referred topic to
        startTimestamp = new Date(), // Date/time when the server started. Used for benchmarking.
        configTM, // The configuration topic map    
        config = {}, // Object containing configuration
        server, // Connect server object
        data, // Temporary variable to hold the topic map
        tmData,
        reader,
        i,
        j,
        topics,
        psis,
        sis,
        parsedIri,
        exportTypes;

    options = [
        { short       : 'c',
            long        : 'config',
            description : 'Filename of the configuration file.',
            value       : true
        },
        { short       : 'l',
            long        : 'logfile',
            description : 'Path and filename to a log file. Must be writable by this process. Defaults to stdout.'
        },
        { short       : 'd',
            long        : 'debug',
            description : 'Sets a debug level. Valid values are \'info\', \'debug\'.',
            value       : true
        },
        { short       : 'p',
            long        : 'port',
            description : 'The port for the web server. The default is 8000'
        },
        { long        : 'list',
            description : 'Lists all PSIs that belong to "servername" and exits',
            value       : false
        },
        {
            long        : 'dc-created',
            description : 'Default create date for PSIs in ISO format (YYYY-MM-DD)',
            value       : true
        },
        {
            long        : 'dc-publisher',
            description : 'Default publisher name for all PSIs',
            value       : true
        },
        {
            long        : 'psi-status',
            description : 'Default publishing status for all PSIs',
            value       : true
        }
    ];

    cmdArguments = [ { name: 'tm', required: true },
        { name: 'servername', required: true } ];
    opts.parse(options, cmdArguments, true);

    servername = config.servername || opts.arg('servername');
    debug = opts.get('d') || 'info';  // default debug value
    tmFile = opts.arg('tm');
    showList = opts.get('list') || false;
    Mu.templateRoot = './templates';

    // Exports a topic as a JTM topic map
    showJTM = function (res, iri, topic, options) {
        var exporttm, writer;
        exporttm = tools.exportTopic(topic.getParent(), topic, options.exportTypes); 
        writer = new tmjs.TM.JTM.Writer(exporttm);
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.write(JSON.stringify(writer.toObject(exporttm)));
        res.end();
        exporttm.remove();
    };

    showHTML = function (res, iri, topic, options) {
        var ctx, type, occs, roles, rolesPlayed, roleTextArr, i,
            j, assoc;
        // TODO: Right now creation-date, publisher and status are hard-coded.
        // They can be set with command line parameters for topic maps
        // that don't have corresponding occurrences, but we're missing
        // some code that tries to fetch these occurrences.

        // Also: showJTM creates a topic map with all occurrences and
        // associations from the configuration topic map. We have to do the
        // same here.

        ctx = {
            psi: iri,
            pathname: options.parsedIri.pathname,
            fields: [
                {label: 'Created', value: config.createdFallback},
                {label: 'Status', value: config.psiStatusFallback}
            ],
            timestamp: function () {
                return (((new Date() - startTimestamp) / 1000) + "secs");
            }
        };
        if (config.publisherFallback) {
            ctx.fields.push({label: 'Publisher', value: opts.get('dc-publisher')});
        }

        // Go through all occurrence and association types that we want to export
        if (options.exportTypes) {
            for (type in options.exportTypes) {
                if (options.exportTypes.hasOwnProperty(type)) {
                    occs = topic.getOccurrences(options.exportTypes[type]);
                    for (j = 0; j < occs.length; j += 1) {
                        ctx.fields.push({label: tools.getTopicLabel(type).label,
                                        value: occs[j].getValue()});
                    }
                }
            }

            // Not the most performant way. Refactor!
            for (type in options.exportTypes) {
                if (options.exportTypes.hasOwnProperty(type)) {
                    rolesPlayed = topic.getRolesPlayed();
                    for (i = 0; i < rolesPlayed.length; i += 1) {
                        assoc = rolesPlayed[i].getParent();
                        if (assoc.getType().equals(options.exportTypes[type])) {
                            roles = assoc.getRoles();
                            // Textual representation of all roles exept the one that topic plays
                            roleTextArr = []; 
                            for (j = 0; j < roles.length; j += 1) {
                                if (!roles[j].equals(rolesPlayed[i])) {
                                    roleTextArr.push(
                                        tools.getTopicLabel(roles[j].getPlayer(),
                                                            {anchor: true}).label + ' (' +
                                        tools.getTopicLabel(roles[j].getType(),
                                                            {anchor: true}).label + ')');
                                }
                            }
                            if (roleTextArr.length > 0) {
                                ctx.fields.push({label: tools.getTopicLabel(assoc.getType(),
                                                                            {anchor: true}).label,
                                                value: roleTextArr.join(', ')});
                            }
                        }
                    }
                }
            }
        }
        ctx.name = tools.getTopicLabel(topic).label;
        res.writeHead(200, { 'Content-Type': 'text/html' });
        Mu.render('psi.html', ctx, {}, function (err, output) {
            if (err) {
                throw err;
            }
            output.addListener('data', function (c) {
                res.write(c);
            }).addListener('end', function () { res.end();
                                            });
        });
    };


    tm = tools.getTopicMap('http://semanticheadache.com/node-psi-server');
    configTM = tools.getTopicMap('http://semanticheadache.com/node-psi-server-config');
    exportIdentifiers = tools.parseConfigTM(configTM, opts, options);
    config.port = config.port || opts.get('p') || '8000';
    config.logFile = config.logFile || opts.get('l') || null;
    config.publisherFallback = config.publisherFallback || opts.get('dc-publisher');
    config.psiStatusFallback = config.psiStatusFallback || opts.get('psi-status') || 'Experimental';
    config.createdFallback = config.createdFallback || opts.get('dc-created') || new Date().toUTCString();
    config.debugLevel = config.debugLevel || opts.get('d') || 'info';
    logStream = null;
    //logStream = fs.open(config.logFile);

    // TODO: stat to check if tmFile exists
    // Use stat's date as default creation date
    data = fs.readFileSync(tmFile);
    tmData = JSON.parse(data);

    reader = new tmjs.TM.JTM.Reader(tm);
    reader.fromObject(tmData);

    util.log("Imported tm in " + ((new Date() - startTimestamp) / 1000) +
             "secs");

    exportTypes = tools.prefetchExportTypes(tm, exportIdentifiers);

    if (showList) {
        // Shows a list of all PSIs that match servername
        // TODO: move this to a function in tools.js
        topics = tm.getTopics();
        util.puts('Servername in showList: ' + servername);
        psis = [];
        util.puts('List of PSIs:');
        for (i = 0; i < topics.length; i += 1) {
            sis = topics[i].getSubjectIdentifiers();
            for (j = 0; j < sis.length; j += 1) {
                parsedIri = url.parse(sis[j].getReference(), true);
                if (parsedIri.host === servername) {
                    psis.push(sis[j].getReference());
                }
            }
        }
        psis.sort();
        for (i = 0; i < psis.length; i += 1) {
            util.puts(' * ' + psis[i]);
        }
        return;
    }

    server = Connect.createServer(
        Connect.logger({stream: logStream}),
        Connect.errorHandler({ showStack: true, dumpExceptions: true }),
        Connect.router(function (app) {
            app.get('/*.css', Connect.staticProvider(__dirname + '/static'));
            app.get('/favicon.ico', Connect.staticProvider(__dirname + '/static'));
        }),
        function (req, res) {
            // Incoming request. Check if a topic with the
            // given PSI exists or serve an error page
            var iri, parsedIri, loc, topic, format;
            parsedIri = url.parse(req.url, true);
            iri = 'http://' + servername + parsedIri.pathname;
            startTimestamp = new Date();
            loc = tm.createLocator(iri);
            topic = tm.getTopicBySubjectIdentifier(loc);

            format = 'xhtml';
            if (parsedIri.query && parsedIri.query.format) {
                if (parsedIri.query.format.match(/^(jtm|xhtml)$/)) {
                    format = parsedIri.query.format;
                }
            }

            if (topic) {
                // TODO:
                // * Find supertype(s) and superclass(es)
                // * Find occurrences from the list of occurrences we want to list
                if (format === 'jtm') {
                    showJTM(res, iri, topic, {exportTypes: exportTypes});
                } else {
                    showHTML(res, iri, topic, {parsedIri: parsedIri,
                             exportTypes: exportTypes});
                }
            } else {
                res.writeHead(404, { 'Content-Type': 'text/html' });
                Mu.render('404.html', {psi: iri,
                          timestamp: function () {
                              return (((new Date() - startTimestamp) / 1000) + "secs");
                          }
                }, {}, function (err, output) {
                    if (err) {
                        throw err;
                    }
                    output.addListener('data', function (c) {
                        res.write(c);
                    }).addListener('end', function () { res.end();
                                                    });
                });
                //throw new Error('oh noes!');
            } 
        }
    ).listen(config.port);
    util.log('Server running at http://127.0.0.1:' + config.port + '/');
}());
